Fix cached clipboard responses during raw transfers

ilammy opened the pull request:
FreeRDP uses `clipboard->data` to cache the result of the Windows→X11 clipboard format conversion, and `xf_cliprdr_process_selection_request()` immediately provides this result to local applications if they request the same clipboard format again. This saves us a possibly costly conversion in case where the user pastes data repeatedly.

However, this caching mechanism did not support raw clipboard transfers (PR #2851) where the unmodified data is passed between two FreeRDP clients. We use the same XClipboard protocol for this, so the `clipboard->data` is in play. We clear the cached value when we receive new data from the server, so initially raw transfers are fine. But if some local application (e.g., a clipboard manager) asks for some data format before the data is pasted into the second FreeRDP session then `clipboard->data` will contain the _converted_ data. And this converted cached data will be provided to the second FreeRDP session as a part of the raw data transfer. Instead we should have provided the original data.

In order to achieve this we are now caching the original data in the same way as the converted one, and the original data is now correctly provided when the second FreeRDP session asks for a raw data transfer.

This should resolve issue #3461.

hardening commented on client/X11/xf_cliprdr.c:
```diff
@@ -1370,9 +1392,21 @@ static UINT xf_cliprdr_server_format_data_response(CliprdrClientContext* context
 
+	/* We have to copy the original data again, as pSrcData is now owned
+	 * by clipboard->system. Memory allocation is not fatal here as this
+	 * is only a cached value. */
```
I think that at least we should print a message.

hardening commented:
@freerdp-bot could you test that please ?

ilammy commented:
I added a warning about the allocation failure into the else branch. It also prints out the amount of memory we have failed to allocate, just in case someone tries to pass `(size_t) -1` there.

bmiklautz commented:
@freerdp-bot test.

bmiklautz commented:
@ilammy could you rebase your PR to the current master?

ilammy commented:
Hi @bmiklautz,

I have rebased the branch onto the current master. I have squashed both commits (the original 10bb24f, and 58d619d with an improvement suggested by @hardening), so it's only one commit now.

The commit 8f1adf64eee1112e2c0942496d7b39841bf1434c ("Refactored ClipboardSetData") introduced some merge conflicts:

- Formatting changes in `xf_cliprdr_process_selection_request()`. No actual conflicts there, just different patch context prevented them from applying cleanly.
- `pSrcData` local variable was removed in `xf_cliprdr_server_format_data_response()`. The original commit was setting it to NULL after freeing. Now we do not need it at all.

I have checked the fix again, it still fixes #3461.

tigerfoot commented:
Ping ?

tigerfoot commented:
@bmiklautz is there anything else than need to be done by @ilammy or one of us ?
I believe this pr would be a nice christmas present :-)

akallabeth commented:
@freerdp-bot test

akallabeth commented:
@bmiklautz Could you have a look why this pull does not finish the checks?

bmiklautz commented:
@freerdp-bot test.

tigerfoot commented:
@bmiklautz @akallabeth seems the bot have sucessfully check the PR.
Now it just need a human review ;-)

tigerfoot commented:
@bmiklautz @akallabeth sorry to bother, but what is missing to get a human review and approval ?

akallabeth commented:
@tigerfoot currently mostly time ;) but this one looks good from the code side.

tigerfoot commented:
At least I'm using that patch to the freeRDP build for openSUSE and has helped to fix 
[boo#1004108](https://bugzilla.opensuse.org/show_bug.cgi?id=1004108)
With report of happy users, I'm also using it since mid-december without any troubles.

bmiklautz commented:
@nfedera any thoughts from your side?

nfedera commented on client/X11/xf_cliprdr.c:
```diff
@@ -1343,8 +1361,25 @@ static UINT xf_cliprdr_server_format_data_response(CliprdrClientContext*
+	}
+	else
+	{
+		WLog_WARN(TAG, "failed to allocate %zu bytes for a copy of raw clipboard data", size);
```
Please see https://github.com/FreeRDP/FreeRDP/blob/master/docs/PrintFormatSpecifiers.md regarding the print format specifiers.

akallabeth commented:
@ilammy If you fix the remark from @nfedera I'll merge this pull

ilammy commented:
Hi. Sorry, I got off track a bit.

`PRIuz` specifier is defined in the upstream WinPR, the base commit of my branch does not contain it. So I have rebased the branch onto the latest master again, and fixed the the format specifier in the warning message.

/cc @nfedera, @akallabeth

akallabeth commented:
@freerdp-bot test

akallabeth commented:
@mfleisz @bmiklautz @nfedera Any objections merging this one?
